<img src="30S.png" width="50%" alt="Image of 30S ribosomal subunit" style="float:right"/>

# Ribosomal Database Project

The Ribosomal Database Project (RDP) aims to provide a comprehensive database of ribosomal RNA (rRNA) sequences and related data for use in phylogenetic analysis and microbial ecology research. The project was initiated in 1984 by Carl Woese, a pioneer in the field of molecular evolution, and is currently maintained by a team of researchers at Michigan State University.

The RDP contains a curated database of rRNA sequences from bacteria, archaea, and eukaryotes. The database includes both 16S and 18S rRNA sequences, which are commonly used as molecular markers for bacterial and eukaryotic phylogenetic analysis, respectively. 

Many bacteria have more than one copy of 16s rRNA genes in their genome, with some having as many as 15 different copies (doi: 10.1371/journal.pone.0057923. Epub 2013 Feb 27. PMID: 23460914)

In this workshop, we'll use a curated file of approximately 15000 16S rDNA sequences.

# Retrieving your Sequences

The 16S sequences obtained by Nanopore sequencing have been provided in a csv file called `class_16S_sequences.csv`.

Samples were identified by using extra sequences known as 'barcodes' that were added to the PCR primers. Each sample was tagged with a unique combination from 8 forward and 15 reverse primer sequences - you noted these down in the lab in week 5 when you set up the PCR. Look up your 2 barcode references, for example A2 or F6. 

# Workshop tasks

Today's workshop is divided into 5 sections:
* Retrieving your 16S sequences
* Loading the reference 16S sequence database
* Searching for database matches
* Visualising your analysis results
* Evaluation and Reflection on Results

### Task 1: Retrieving your 16S sequences

1.1> Load the file 'Discovery_16S_consensus.csv' into a Pandas dataframe and explore the contents. n.b. the 'descriptor' is a pseudo-random identifier for your sequence (your can ignore it). 

1.2> Retrieve your 16S sequences from the Pandas dataframe using the `.query()` function by looking for rows that match your combination of forward and reverse primers. If there is no match for your group's 2 pairs of forward and reverse primers, run the code below to show which combinations of primers do have reads, then and and select one you like the look of for further analysis in this workshop. 
```python 
#load the dataframe
df=pd.read_csv("class_16S_sequences.csv")
# Create a contingency table of the forward and reverse indices
ct = pd.crosstab(df['reverse_primer'], df['forward_primer'])
print(ct)
#Add code below this that uses .query to filter the results for your combination of forward and reverse primers. 
```

1.3> Extract your 16S sequences to a list using `.tolist()`. From the list, extracting individual sequences using indexing (for example `my_sequences[0]` or `my_sequences[1]`).

```python
seq_list=my_sequences['consensus_sequence'].tolist()
seq_list[0]
```


In [2]:
import pandas as pd

df=pd.read_csv("class_16S_sequences.csv")
ct = pd.crosstab(df['reverse_primer'], df['forward_primer'])
print(ct)

forward_primer  A  B  C  D  E  F  G  H  I  J  K  L  U  V  Y
reverse_primer                                             
0               1  0  3  1  0  0  0  0  1  0  1  1  0  0  0
1               0  0  0  0  1  1  0  0  0  0  0  0  0  0  0
2               0  1  0  0  1  0  0  0  2  1  0  0  0  0  0
3               0  0  1  0  0  0  0  1  0  1  0  1  0  0  0
4               0  0  0  0  0  1  0  1  0  0  0  1  0  0  0
5               0  0  1  0  0  0  0  0  0  0  0  2  0  0  0
6               1  0  0  0  0  1  0  0  0  0  0  0  0  0  0
7               0  1  0  0  0  1  0  0  1  0  0  0  0  0  1
8               0  1  0  0  0  0  0  0  0  0  0  0  0  0  0
9               1  0  0  0  2  0  0  0  2  2  0  0  0  1  0
10              0  1  0  0  0  0  0  0  1  0  0  1  0  1  0
11              0  0  1  0  0  1  0  0  0  0  0  1  0  0  0
12              0  1  1  0  0  0  0  0  0  0  0  0  1  0  0
13              0  1  0  2  0  0  2  3  0  0  1  0  0  0  0
14              1  2  2  1  0  0  0  0  

In [2]:
import pandas as pd

df=pd.read_csv("class_16S_sequences.csv")
ct = pd.crosstab(df['reverse_primer'], df['forward_primer'])
print(ct)
consensus_sequence=df.query("forward_primer== 'E' and reverse_primer== 9")
seq_list= consensus_sequence['consensus_sequence'].tolist()
display(seq_list[0])
display(consensus_sequence)

# Display both sequences if there are more than one
if len(seq_list) > 0:
    display(seq_list[0])  # First sequence
if len(seq_list) > 1:
    display(seq_list[1])  # Second sequence

# Display the filtered consensus_sequence DataFrame
display(consensus_sequence)

#Your code here

forward_primer  A  B  C  D  E  F  G  H  I  J  K  L  U  V  Y
reverse_primer                                             
0               1  0  3  1  0  0  0  0  1  0  1  1  0  0  0
1               0  0  0  0  1  1  0  0  0  0  0  0  0  0  0
2               0  1  0  0  1  0  0  0  2  1  0  0  0  0  0
3               0  0  1  0  0  0  0  1  0  1  0  1  0  0  0
4               0  0  0  0  0  1  0  1  0  0  0  1  0  0  0
5               0  0  1  0  0  0  0  0  0  0  0  2  0  0  0
6               1  0  0  0  0  1  0  0  0  0  0  0  0  0  0
7               0  1  0  0  0  1  0  0  1  0  0  0  0  0  1
8               0  1  0  0  0  0  0  0  0  0  0  0  0  0  0
9               1  0  0  0  2  0  0  0  2  2  0  0  0  1  0
10              0  1  0  0  0  0  0  0  1  0  0  1  0  1  0
11              0  0  1  0  0  1  0  0  0  0  0  1  0  0  0
12              0  1  1  0  0  0  0  0  0  0  0  0  1  0  0
13              0  1  0  2  0  0  2  3  0  0  1  0  0  0  0
14              1  2  2  1  0  0  0  0  

'GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACCGACGGGAGCTTGCTCCCTTAGGTCAGCGGCGGACGGGTGAGTAACACGTGGGTAACCTGCCTGTAAGACTGGGATAACTCCGGGAAACCGGGGCTAATACCGGATGCTTGATTGAACCGCATGGTTCAATCATAAAAGGTGGCTTTTAGCTACCACTTACAGATGGACCCGCGGCGCATTAGCTAGTTGGTGAGGTAACGGCTCACCAAGGCGACGATGCGTAGCCGACCTGAGAGGGTGATCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTAGGGAATCTTCCGCAATGGACGAAAGTCTGACGGAGCAACGCCGCGTGAGTGATGAAGGTTTTCGGATCGTAAAACTCTGTTGTTAGGGAAGAACAAGTACCGTTCGAATAGGGCGGTACCTTGACGGTACCTAACCAGAAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCGCGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCCCGGCTCAACCGGGGAGGGTCATTGGAAACTGGGGAACTTGAGTGCAGAAGAGGAGAGTGGAATTCCACGTGTAGCGGTGAAATGCGTAGAGATGTGGAGGAACACCAGTGGCGAAGGCGACTCTCTGGTCTGTAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCGAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAGTGCTAAGTGTTAGAGGGTTTCCGCCCTTTAGTGCTGCAGCAAACGCATTAAGCACTCCGCCTGGGGAGTACGGTCGCAAGACTGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGAGCATGTGGTTTAATTCGAAGCAACGCGAAGAACCTTACCAGGTCTTGACATCCTCTGACAACCCTAGAGAT

forward_primer  reverse_primer              descriptor  \
27              E               9  consensus_E_9_0_0(393)   
28              E               9  consensus_E_9_0_1(111)   

                                   consensus_sequence  16S_seq_length  
27  GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACCG...            1541  
28  ATTGAACGCTAGCGGGATGCTTTACACATGCAAGTCGAACGGCAGC...            1531

'GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACCGACGGGAGCTTGCTCCCTTAGGTCAGCGGCGGACGGGTGAGTAACACGTGGGTAACCTGCCTGTAAGACTGGGATAACTCCGGGAAACCGGGGCTAATACCGGATGCTTGATTGAACCGCATGGTTCAATCATAAAAGGTGGCTTTTAGCTACCACTTACAGATGGACCCGCGGCGCATTAGCTAGTTGGTGAGGTAACGGCTCACCAAGGCGACGATGCGTAGCCGACCTGAGAGGGTGATCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTAGGGAATCTTCCGCAATGGACGAAAGTCTGACGGAGCAACGCCGCGTGAGTGATGAAGGTTTTCGGATCGTAAAACTCTGTTGTTAGGGAAGAACAAGTACCGTTCGAATAGGGCGGTACCTTGACGGTACCTAACCAGAAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCGCGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCCCGGCTCAACCGGGGAGGGTCATTGGAAACTGGGGAACTTGAGTGCAGAAGAGGAGAGTGGAATTCCACGTGTAGCGGTGAAATGCGTAGAGATGTGGAGGAACACCAGTGGCGAAGGCGACTCTCTGGTCTGTAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCGAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGAGTGCTAAGTGTTAGAGGGTTTCCGCCCTTTAGTGCTGCAGCAAACGCATTAAGCACTCCGCCTGGGGAGTACGGTCGCAAGACTGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGAGCATGTGGTTTAATTCGAAGCAACGCGAAGAACCTTACCAGGTCTTGACATCCTCTGACAACCCTAGAGAT

'ATTGAACGCTAGCGGGATGCTTTACACATGCAAGTCGAACGGCAGCGCGAAGAAAGCTTGCTTTCTTTGGCGGCGAGTGGCGAACGGGTGAGTAATATATCGGAACGTGCCCAGTAGCGGGGGATAACTACGCGAAAGCGTGGCTAATACCGCATACGCCCTACGGGGGAAAGGGGGGGATCGCAAGACCTCTCACTATTGGAGCGGCCGATACCAGATTAGCTTGTTGGTGAGGTAAAGGCTCACCAAGGCCGCGATCTGTAGCTGGTTTGAGAGGACGACCAGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTGGGGAATTTTGGACAATGGGGGAAACCCTGATCCAGCCATCCCGCGTGTGCGATGAAGGCCTTCGGGTTGTAAAGCACTTTTGGCAGGGAAGAAAAGGCGCCGGATAATACCTGGCGCTGATGACGGTACCTGCAGAATAAGCACCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGTTCGGAAAGAAAGATGTGAAATCCCAGAGCTTAACTTTGGAACTGCATTTTTAACTGCCGAGCTAGAGTATGTCAGAGGGGGGTAGAATTCCACGTGTAGCAGTGAAATGCGTAGAGATGTGGAGGAATACCGATGGCGAAGGCAGCCCCCTGGGATAATACTGACGCTCATGCACGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCCTAAACGATGTCAACTAGCTGTTGGGGCCTTCGGGCCTTAGTAGCGCAGCTAACGCGTGAAGTTGACCGCCTGGGGAGTACGGTCGCAAGATTAAAACTCAAAGGAATTGACGGGGACCCGCACAAGCGGTGGATGATGTGGATTAATTCGATGCAACGCGAAAAACCTTACCTACCCTTGACATGTCTGGAAGCTTCGAGAGATTGAAGTGT

forward_primer  reverse_primer              descriptor  \
27              E               9  consensus_E_9_0_0(393)   
28              E               9  consensus_E_9_0_1(111)   

                                   consensus_sequence  16S_seq_length  
27  GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACCG...            1541  
28  ATTGAACGCTAGCGGGATGCTTTACACATGCAAGTCGAACGGCAGC...            1531

### Task 2: Loading the reference 16S sequence database

2.1> Load the 16S_reference.csv database file into a Pandas dataframe (use a different dataframe name to that you used when loading your sequences). This is a large version of the 'mini' dataframe you used last week. 

2.2> How many sequences are in the database?




In [11]:
#Your code here
df1=pd.read_csv('16S_reference.csv')

print(len(df1))
display(df1)

15155


id  \
0      Bacteria;Actinobacteria;Actinobacteria;Actinom...   
1      Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   
2      Bacteria;Actinobacteria;Actinobacteria;Actinom...   
3      Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   
4      Bacteria;Proteobacteria;Betaproteobacteria;Bur...   
...                                                  ...   
15150  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
15151  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
15152  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
15153  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
15154  Bacteria;Proteobacteria;Gammaproteobacteria;En...   

                                                sequence  \
0      GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   
1      GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   
2      TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   
3      AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   
4      ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   
...                                                  ...   
15150  AGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACACA...   
15151  GATGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   
15152  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAA...   
15153  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAA...   
15154  ACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGTAGCACAGa...   

                                      species  
0      Mycobacterium_heidelbergense(AJ000684)  
1                 Vibrio_atlanticus(EF599163)  
2          Mycobacterium_aubagnense(AY859683)  
3          Acetobacter_senegalensis(AY883036)  
4        Aquincola_tertiaricarbonis(DQ656489)  
...                                       ...  
15150      Leucobacter_komagatae(NR_112021.1)  
15151         Thermobifida_fusca(NR_112015.1)  
15152     Microbispora_mesophila(NR_112014.1)  
15153     Parvopolyspora_pallida(NR_112013.1)  
15154         Klebsiella_oxytoca(NR_112010.1)  

[15155 rows x 3 columns]


### Task 3: Searching for database matches

3.1> For each query 16S sequence (i.e. for each of your own unknown 16S sequences), create a new column in the Pandas dataframe that represents the Levenshtein distance between that sequence and the sequences in the dataframe. The code you developed at the end of the week 7 workshop is close to what you will need to use here. Make sure that you create a different column name for the distance to each 16S sequence youre comparing to, for example 'dist_to_seq_1', 'dist_to_seq_2', etc. You can do this by run each of your group's sequences in turn, or each person can analyse a different one. If you're looking for an additional challenge, you could use a loop to process all sequences in your list of your group's sequences.

3.2> Identify the best match(es) for each sequence: how close is the match? Hint: to find the closest match, you'll need to identify which row has the minimum value in the respective distance column. You then need to use .query() to return the row with that value.  You can adapt the code you developed in the Week 8 exercise notebook to do this. 


In [8]:
!pip install levenshtein
#pip is a package manager:this command installs the levenshtein package on our Noteable instance.

  Using cached levenshtein-0.27.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (160 kB)
  Using cached rapidfuzz-3.12.2-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)
  Attempting uninstall: rapidfuzz
    Found existing installation: rapidfuzz 3.2.0
    Uninstalling rapidfuzz-3.2.0:
      Successfully uninstalled rapidfuzz-3.2.0


In [12]:
df1.head(10)

id  \
0  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
1  Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   
2  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
3  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   
4  Bacteria;Proteobacteria;Betaproteobacteria;Bur...   
5  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
6  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   
7  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
8  Bacteria;Proteobacteria;Gammaproteobacteria;Al...   
9  Bacteria;Actinobacteria;Actinobacteria;Actinom...   

                                            sequence  \
0  GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   
1  GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   
2  TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   
3  AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   
4  ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   
5  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...   
6  ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...   
7  TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...   
8  AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...   
9  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   

                                  species  
0  Mycobacterium_heidelbergense(AJ000684)  
1             Vibrio_atlanticus(EF599163)  
2      Mycobacterium_aubagnense(AY859683)  
3      Acetobacter_senegalensis(AY883036)  
4    Aquincola_tertiaricarbonis(DQ656489)  
5         Couchioplanes_caeruleus(D85479)  
6   Maritimibacter_alkaliphilus(DQ915443)  
7        Nocardioides_fonticola(EF626689)  
8       Shewanella_schlegeliana(AB081760)  
9                Lentzea_albida(AB006176)

In [13]:
from Levenshtein import distance
string_to_compare = seq_list[0]
df1['distance'] = df1['sequence'].apply(distance, args=(string_to_compare,))

display(df1.head(10))

id  \
0  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
1  Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   
2  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
3  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   
4  Bacteria;Proteobacteria;Betaproteobacteria;Bur...   
5  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
6  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   
7  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
8  Bacteria;Proteobacteria;Gammaproteobacteria;Al...   
9  Bacteria;Actinobacteria;Actinobacteria;Actinom...   

                                            sequence  \
0  GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   
1  GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   
2  TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   
3  AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   
4  ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   
5  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...   
6  ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...   
7  TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...   
8  AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...   
9  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   

                                  species  distance  
0  Mycobacterium_heidelbergense(AJ000684)       387  
1             Vibrio_atlanticus(EF599163)       403  
2      Mycobacterium_aubagnense(AY859683)       388  
3      Acetobacter_senegalensis(AY883036)       473  
4    Aquincola_tertiaricarbonis(DQ656489)       412  
5         Couchioplanes_caeruleus(D85479)       364  
6   Maritimibacter_alkaliphilus(DQ915443)       434  
7        Nocardioides_fonticola(EF626689)       415  
8       Shewanella_schlegeliana(AB081760)       417  
9                Lentzea_albida(AB006176)       345

In [21]:
print(df1.columns)
# Find the minimum distance value
min_distance = df1['distance'].min()

# Return the row with that minimum distance
best_match = df1.query("distance == @min_distance")

display(best_match)



Index(['id', 'sequence', 'species', 'distance'], dtype='object')


id  \
9440  Bacteria;Firmicutes;Bacilli;Bacillales;Bacilla...   

                                               sequence  \
9440  GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACAG...   

                            species  distance  
9440  Bacillus_atrophaeus(AB021181)        80

# Task 4: Visualising your analysis results

4.1> Run a needleman_wunsch alignment between your 16S sequence(s) and the best match. Python code to run an Needleman Wunsch alignment is shown below. You'll need to extract your own 16S sequence and the best match 16S sequence as strings before you can align them. 
4.2> View the alignment. How well is the sequence aligned? Can you see poorer alignment at the ends of the sequence caused by the presence of PCR primers in your sequence?

Hint: If you're struggling with getting the sequence to convert to a string and you are seeing the text 'Name: sequence, dtype: object' when you print it, try suffixing the expression with .iloc[0]. This extracts the string from the Pandas data series.

An example code for running a Needleman-Wunsch alignment is shown below:

```python
import needleman_wunsch
seq1='ATGCTGAGCTAGCGGCTATATTCTATCGGGAGCGATTTACTACTC'
seq2='ATGCTAGGTAGCGGACTATATACTATCGCGAGCGATTAACTAGCC'
print(needleman_wunsch.align(seq1, seq2))
```
Expected output:
```
seq1: ATGCTGAGCTAGCGG-CTATATTCTATCGGGAGCGATTTACTA-CTC
      ||||| || |||||| |||||| |||||| |||||||| |||| | |
seq2: ATGCT-AGGTAGCGGACTATATACTATCGCGAGCGATTAACTAGC-C
```



In [29]:
import needleman_wunsch
best_match_sequence = best_match['sequence'].iloc[0]

# Display the alignment results for both
print(needleman_wunsch.align(seq_list[0], best_match_sequence))

#Your code here

seq1: GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACCGACGGGAGCTTGCTCCCTTAGGTCA
      |||||||||||||||||||||||||||||||||||||||||||| || ||||||||||||||| | || |
seq2: GACGAACGCTGGCGGCGTGCCTAATACATGCAAGTCGAGCGGACAGATGGGAGCTTGCTCCCTGATGTTA

seq1: GCGGCGGACGGGTGAGTAACACGTGGGTAACCTGCCTGTAAGACTGGGATAACTCCGGGAAACCGGGGCT
      ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
seq2: GCGGCGGACGGGTGAGTAACACGTGGGTAACCTGCCTGTAAGACTGGGATAACTCCGGGAAACCGGGGCT

seq1: AATACCGGATGCTTGATTGAACCGCATGGTTCAATCATAAAAGGTGGCTTTTAGCTACCACTTACAGATG
      ||||||||||||||| |||||||||||||||||| ||||||||||||| ||  |||||||||||||||||
seq2: AATACCGGATGCTTGTTTGAACCGCATGGTTCAAACATAAAAGGTGGC-TTCGGCTACCACTTACAGATG

seq1: GACCCGCGGCGCATTAGCTAGTTGGTGAGGTAACGGCTCACCAAGGCGACGATGCGTAGCCGACCTGAGA
      ||||||||||||||||||||||||||||||||||||||||||||||| ||||||||||||||||||||||
seq2: GACCCGCGGCGCATTAGCTAGTTGGTGAGGTAACGGCTCACCAAGGCAACGATGCGTAGCCGACCTGAGA

seq1: GGGTGATCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTAGGGAATC

In [30]:
print(needleman_wunsch.align(seq_list[1], best_match_sequence))

seq1: -ATTGAACGCTAGCGG-GATGCTTTACACATGCAAGTCGAACGGCAGCGCGAAGAAAGCTTGCTTTCTTT
       |  ||||||| |||| | ||| | | ||||||||||||| ||| | |  || |  |||||||  ||  |
seq2: GA-CGAACGCTGGCGGCG-TGCCTAATACATGCAAGTCGAGCGG-A-C-AGATGGGAGCTTGC--TCCCT

seq1: GGCGGCGAGTGGCGAACGGGTGAGTAATATATCGG-AACGTGCCCAGT-AG-CGGGGGATAACTACGCGA
       |  |  || |||| |||||||||||| |  | || ||| || || || || |  ||||||||| || ||
seq2: -GATGTTAGCGGCGGACGGGTGAGTAACACGTGGGTAACCTG-CCTGTAAGAC-TGGGATAACTCCGGGA

seq1: AAGCGTGGCTAATACCGCATAC------G--CC-CTACGG------GGGAAAGGGGGGGATC-GCAAGAC
      || || ||||||||||| || |      |  || | | ||         ||| || ||  || ||   ||
seq2: AACCGGGGCTAATACCGGATGCTTGTTTGAACCGC-ATGGTTCAAACATAAAAGGTGGCTTCGGC--TAC

seq1: CTCTCACTA-TTGGA---GCGGCCGATACCAGATTAGCTTGTTGGTGAGGTAAAGGCTCACCAAGGCCGC
      | || || |  ||||   |||| ||    |  ||||||| ||||||||||||| |||||||||||||  |
seq2: CACTTAC-AGATGGACCCGCGG-CG----C--ATTAGCTAGTTGGTGAGGTAACGGCTCACCAAGGCAAC

seq1: GAT-CTGTAGCTGGTTTGAGAGGACGACCAGCCACACTGGGACTGAGACACGGCCCAGACTCCTAC

# Finding antibiotic producers from soil sample

Using the same method shown in Task 1 to 4, the species of antiobiotic-producing bacteria found from my soil sample were identified. The sample was collected from Almond River, Edinburgh at coordinates 55°57'07.6"N, 3°24'12.7"W. The species' 16S rDNAs sequences were retrieved from PCR test conducted at Edinburgh University's laboratory. These sequences were then compared against the sequences in Ribosome Database Project. 



In [1]:
import pandas as pd

df=pd.read_csv("class_16S_sequences_week_9.csv")
ct = pd.crosstab(df['reverse_primer'], df['forward_primer'])
print(ct)
consensus_sequence = df.query("(forward_primer == 'C' and reverse_primer == 1) | (forward_primer == 'J' and reverse_primer == 5)| (forward_primer == 'E' and reverse_primer == 11)| (forward_primer == 'J' and reverse_primer == 18)")
seq_list= consensus_sequence['consensus_sequence'].tolist()

# Display both sequences if there are more than one
if len(seq_list) > 0:
    display(seq_list[0])  # First sequence
if len(seq_list) > 1:
    display(seq_list[1])  # Second sequence
if len(seq_list) > 2:
    display(seq_list[2]) 
if len(seq_list) > 3:
    display(seq_list[3]) 
# Display the filtered consensus_sequence DataFrame
display(consensus_sequence)

#Your code here
#c1=16 (seqlist0),e11=15(seqlist1), j5=c6(seqlist2)

forward_primer  A  B  C  D  E  F  G  H  I  J  L  R  S  U  V  W  Z
reverse_primer                                                   
0               0  1  2  0  0  0  0  0  1  0  0  0  0  0  0  0  0
1               0  0  1  1  0  1  0  1  0  0  0  0  0  0  0  0  0
2               0  3  0  1  0  0  0  0  0  2  1  0  0  0  0  0  0
3               0  0  2  1  1  1  1  0  0  3  0  0  0  0  0  0  1
4               1  0  0  3  0  0  1  0  1  1  0  0  0  0  0  0  0
5               0  0  1  0  0  0  0  0  2  1  0  0  0  0  0  0  0
6               0  1  0  0  1  0  0  1  0  0  0  0  0  1  0  0  0
7               1  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0
8               1  2  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0
9               0  2  0  0  0  0  1  0  0  1  0  0  0  0  0  0  0
10              0  0  0  0  0  0  1  0  0  2  0  1  1  0  1  0  0
11              0  1  0  0  1  0  1  0  1  0  0  0  0  0  0  0  0
12              0  0  0  0  0  0  0  0  0  2  0  0  0  0  0  0  0
16        

'TAAACCTCGTCGCGTGCTTAACACATGCAAGTCGAACGGTGAAGCCCAGCTTGCTGGGTGGATCAGTGGCGAACGGGTGAGTAACACGTGAGCAACCTGCCCCTGACTCTGGGATAAGCGCTGGAAACGGCGTCTAATACTGGATATGTGACGTGACCGCATGGTCTGCGTCTGAAAGAATTTCGGTTGGGGATGGGCTCGCGGCCTATCAGCTTGTTGGTGAGGTAATGGCTCACCAAGGCGTCGACGGGTAGCCGGCCTGAGAGGGTGACCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGCACAATGGGCGCAAGCCTGATGCAGCAACGCCGCGTGAGGGATGACGGCCTTCGGGTTGTAAACCTCTTTTAGCAGGGAAGAAGCGAAAGTGACGGTACCTGCAGAAAAAGCGCCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGCGCAAGCGTTATCCGGAATTATTCGGCATAAAGTCGTAGGCGGTTTGTCGCGTCTGCTGTGAAATCCGGAGGCTCAACCTCCGGCCTGCAGTGGGTACGGGCAGACTAGAGTGCGGTAGGGGAGTCAAATCCTGGTGTAGCGGTGGAATGCGCAGATATCAGGAGGAACACCGATGGCGAAGGCAGATCTCTGGGCCGTAACTGACGCTGAGGAGCGAAAGGGTGGGGAGCGGACGACTATACCCCTGGTAGTCCACCCCGTAAACGTTGGGAACTAGTTGTGGGGTCCATTCCACGGATTCCGTGACGCAGCTAACGCATTAAGACTCCCGCCTGGGGAGTACGGCCGCAAGGCTAAAACTCAAAGGAATTGACGGGGACCCGCACAAGCGGCGGAGCATGCGGATTAATTCGATGCAACGCGAAGAACCTTACCAAGGCTTGACATATACCGGAAAGCCGTAGAGATACGGCCCCCCTTGTGGTCGGTATACAGGTGGTGCATGGCTGTC

'GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGATGAACCACTTCGGTGGGGATTAGTGGCGAACGGGTGAGTAACACGTGGGCAATCTGCCCTGCACTCTGGGACAAGCCCTGGAAACGGGGTCTAATACCGGATACTGACCCGCTTGGGCATCCAAGCGGTTCGAAAGCTCCGGCGGTGCAGGATGAGCCCGCGGCCTATCAGCTTGTTGGTGAGGTAATGGCTCACCAAGGCGACGACGGGTAGCCGGCCTGAGAGGGCGACCGGCCACACTGGGACTGAGACACGGCCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGCACAATGGGCGAAAGCCTGATGCAGCGACGCCGCGTGAGGGATGACGGCCTTCGGGTTGTAAACCTCTTTCAGCAGGGAAGAAGCGAAAGTGACGGTACCTGCAGAAGAAGCGCCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGAGCTCGTAGGCGGCTTGTCACGTCGGTTGTGAAAGCCCGGGGCTTAACCCCGGGTCTGCAGTCGATACGGGCAGGCTAGAGTTCGGTAGGGGAGATCGGAATTCCTGGTGTAGCGGTGAAATGCGCAGATATCAGGAGGAACACCGGTGGCGAAGGCGGATCTCTGGGCCGATACTGACGCTGAGGAGCGAAAGCGTGGGGAGCGAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGGTGGGCACTAGGTGTGGGCGACATTCCACGTCGTCCGTGCCGCAGCTAACGCATTAAGTGCCCCGCCTGGGGAGTACGGCCGCAAGGCTAAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGCGGAGCATGTGGCTTAATTCGACGCAACGCGAAGAACCTTACCAAGGCTTGACATACACCGGAAAGCATCAGAGATGGTGCCCCCCTTGTGGTCGGTGTACAGGTGGTG

'AGTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGCAGCACAGGGGAGCTTGCTCCCTGGGTGGCGAGTGGCGGACGGGTGAGGAATACATCGGAATCTACCTTGTCGTGGGGGATAACGTAGGGAAACTTACGCTAATACCGCATACGACCTACGGGTGAAAGCGGAGGACCGCAAGGCTTCGCGCGATTGGATGAGCCGATGTCGGATTAGCTTGTTGGCGGGGTAACGGCCCACCAAGGCGACGATCCGTAGCTGGTCTGAGAGGATGATCAGCCACACTGGAACTGAGACACGGTCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGGACAATGGGCGCAAGCCTGATCCAGCCATGCCGCGTGAGTGAAGAAGGCCCTCGGGTTGTAAAGCTCTTTTGTTGGGGAAGAAAAGCAATCGGTTAATACCCGGTTGTCCTGACGGTACCCAAAGAATAAGCACCGGCTAACTTCGTGCCAGCAGCCGCGGTAATACGAAGGGTGCAAGCGTTACTCGGAATTACTGGGCGTAAAGCGTGCGTAGGTGGTTTGTTAAGTCTGATGTGAAAGCCCTGGGCTCAACCTGGGAATTGCACTGGATACTGGCAGGCTAGAGTGCGGTAGAGGATGGCGGAATTCCCGGTGTAGCAGTGAAATGCGTAGAGATCGGGAGGAACATCTGTGGCGAAGGCGGCCATCTGGACCAGCACTGACACTGAGGCACGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCCTAAACGATGCGAACTGGATGTTGGGAGCAACTAGGCTCTCAGTATCGAAGCTAACGCGTTAAGTTCGCCGCCTGGGGAGTACGGTCGCAAGACTGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGAGTATGTGGTTTAATTCGATGCAACGCGAAGAACCTTACCTGGCCTTGACATCCACGGAACTTACCAGAGATGGTTTGGT

'ATTGAACGCTGGCGGCAGGCCTGACACCGTGCAGATCAGACAATCAGAAGAGCTTGCTCTTCAATTCAGCGGCGGACGGGTGAGTAATGCCTAGGAATCTGCCTGGTAGTGGGGGACAACGTTTCGAAAGGAACGCTAATACCGCATACGTCCTACGGGAGAAAGCAGGGAGCCTCGGGCCTTGCGCTATCAGATGAGCCTAGGTCGGATTAGCTAGTTGGTGAGGTAATGGCTCACCAAGGCGACGATCCGTAACTGGTCTGAGAGGATGATCAGTCACACTGGAACTGAGACACGGTCCAGACTCCTACGGGAGGCAGCAGTGGGGAATCTGGACAATGGGCGAAAGCCTAGTCCAGCCATGCCGCGTGTGTGAAGAAGGTCTTCGGATTGTAAAGCACTTTAAGTTGGGAGGAGGGCATGCACCTAATACGTGTGTGTCTTGACGTTACCGACAGAATAAGCACCGGCTAACTCTGTGCCAGCAGCCGCGGTAATACAGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGGTGGTTCGTTGAGTTGAATGTGAAATCCCCGGGCTCAACCTGGGAACTGCATCAAAACTGGCAAGCTGAGTTCTGGTAGAGGGTGGTGGAATTTCCTGTGTAGCGGTGAGATGATAGGAACACCAGTGGCGAAGGCGACCACCTGGACTGATACTGACACTGAGGTGCGAGAGCGGGAAGCGAAGCGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGTCAACTAGCCGTTGGGAGCCTTGAGCTCTTAGTGGCGCAGCTAACGCATTAAGTTGACCGCACGCCTGGGGATGCGGCCGCAAGGTTAAAACTCAAATGAATTGACGGGGGCCCGCACAAGCGGTGGAGCATGTCGGATTAGATTGGAGGACAACGCGAAGAACCTTGCCTTGACATCAATGAACTTTCCAGAGATGGATTGGTGCCTTCGGGAACATTGAGACAGG

forward_primer  reverse_primer                            descriptor  \
17              C               1  54ad1885-9646-4c70-8619-de9ae41c8821   
32              E              11                consensus_E_11_0_0(61)   
59              J               5                 consensus_J_5_0_0(24)   
67              J              18  0064a29b-3204-48b7-ae7b-8cfd0834c1d3   
68              J              18  f963612b-fe12-4737-8f80-928f3ed57e0f   

                                   consensus_sequence  16S_seq_length  
17  TAAACCTCGTCGCGTGCTTAACACATGCAAGTCGAACGGTGAAGCC...            1469  
32  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGATGAA...            1500  
59  AGTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGCAGC...            1509  
67  ATTGAACGCTGGCGGCAGGCCTGACACCGTGCAGATCAGACAATCA...            1468  
68  ATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGATGA...            1501

In [3]:
df1=pd.read_csv('16S_reference-Copy1.csv')
display(df1)

id  \
0      Bacteria;Actinobacteria;Actinobacteria;Actinom...   
1      Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   
2      Bacteria;Actinobacteria;Actinobacteria;Actinom...   
3      Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   
4      Bacteria;Proteobacteria;Betaproteobacteria;Bur...   
...                                                  ...   
15150  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
15151  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
15152  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
15153  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
15154  Bacteria;Proteobacteria;Gammaproteobacteria;En...   

                                                sequence  \
0      GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   
1      GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   
2      TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   
3      AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   
4      ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   
...                                                  ...   
15150  AGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACACA...   
15151  GATGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   
15152  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAA...   
15153  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAA...   
15154  ACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGTAGCACAGa...   

                                      species  
0      Mycobacterium_heidelbergense(AJ000684)  
1                 Vibrio_atlanticus(EF599163)  
2          Mycobacterium_aubagnense(AY859683)  
3          Acetobacter_senegalensis(AY883036)  
4        Aquincola_tertiaricarbonis(DQ656489)  
...                                       ...  
15150      Leucobacter_komagatae(NR_112021.1)  
15151         Thermobifida_fusca(NR_112015.1)  
15152     Microbispora_mesophila(NR_112014.1)  
15153     Parvopolyspora_pallida(NR_112013.1)  
15154         Klebsiella_oxytoca(NR_112010.1)  

[15155 rows x 3 columns]

In [4]:
!pip install levenshtein
#pip is a package manager:this command installs the levenshtein package on our Noteable instance.

  Using cached levenshtein-0.27.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (160 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 4.1 MB/s eta 0:00:0000:0100:01
  Attempting uninstall: rapidfuzz
    Found existing installation: rapidfuzz 3.2.0
    Uninstalling rapidfuzz-3.2.0:
      Successfully uninstalled rapidfuzz-3.2.0


In [5]:
from Levenshtein import distance

# Compare to seq_list[0]
string_to_compare_0 = seq_list[0]
df1['distance_0'] = df1['sequence'].apply(distance, args=(string_to_compare_0,))

# Display the updated DataFrame with distance_0
display(df1.head(10))

# Find the minimum distance for distance_0
min_distance_0 = df1['distance_0'].min()
best_match_0 = df1.query("distance_0 == @min_distance_0")
display(best_match_0)

# Compare to seq_list[1]
string_to_compare_1 = seq_list[1]
df1['distance_1'] = df1['sequence'].apply(distance, args=(string_to_compare_1,))

# Display the updated DataFrame with distance_1
display(df1.head(10))

# Find the minimum distance for distance_1
min_distance_1 = df1['distance_1'].min()
best_match_1 = df1.query("distance_1 == @min_distance_1")
display(best_match_1)

# Compare to seq_list[2]
string_to_compare_2 = seq_list[2]
df1['distance_2'] = df1['sequence'].apply(distance, args=(string_to_compare_2,))

# Display the updated DataFrame with distance_2
display(df1.head(10))

# Find the minimum distance for distance_2
min_distance_2 = df1['distance_2'].min()
best_match_2 = df1.query("distance_2 == @min_distance_2")
display(best_match_2)


id  \
0  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
1  Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   
2  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
3  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   
4  Bacteria;Proteobacteria;Betaproteobacteria;Bur...   
5  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
6  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   
7  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
8  Bacteria;Proteobacteria;Gammaproteobacteria;Al...   
9  Bacteria;Actinobacteria;Actinobacteria;Actinom...   

                                            sequence  \
0  GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   
1  GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   
2  TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   
3  AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   
4  ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   
5  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...   
6  ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...   
7  TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...   
8  AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...   
9  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   

                                  species  distance_0  
0  Mycobacterium_heidelbergense(AJ000684)         244  
1             Vibrio_atlanticus(EF599163)         447  
2      Mycobacterium_aubagnense(AY859683)         239  
3      Acetobacter_senegalensis(AY883036)         458  
4    Aquincola_tertiaricarbonis(DQ656489)         420  
5         Couchioplanes_caeruleus(D85479)         219  
6   Maritimibacter_alkaliphilus(DQ915443)         433  
7        Nocardioides_fonticola(EF626689)         273  
8       Shewanella_schlegeliana(AB081760)         423  
9                Lentzea_albida(AB006176)         214

id  \
8329  Bacteria;Actinobacteria;Actinobacteria;Actinom...   

                                               sequence  \
8329  TGGCGGCGTGCTTAACACATGCAAGTCGAACGGTGAACACGGAGCT...   

                             species  distance_0  
8329  Microbacterium_oxydans(Y17227)         140

id  \
0  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
1  Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   
2  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
3  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   
4  Bacteria;Proteobacteria;Betaproteobacteria;Bur...   
5  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
6  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   
7  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
8  Bacteria;Proteobacteria;Gammaproteobacteria;Al...   
9  Bacteria;Actinobacteria;Actinobacteria;Actinom...   

                                            sequence  \
0  GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   
1  GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   
2  TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   
3  AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   
4  ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   
5  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...   
6  ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...   
7  TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...   
8  AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...   
9  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   

                                  species  distance_0  distance_1  
0  Mycobacterium_heidelbergense(AJ000684)         244         228  
1             Vibrio_atlanticus(EF599163)         447         411  
2      Mycobacterium_aubagnense(AY859683)         239         226  
3      Acetobacter_senegalensis(AY883036)         458         435  
4    Aquincola_tertiaricarbonis(DQ656489)         420         409  
5         Couchioplanes_caeruleus(D85479)         219         182  
6   Maritimibacter_alkaliphilus(DQ915443)         433         400  
7        Nocardioides_fonticola(EF626689)         273         256  
8       Shewanella_schlegeliana(AB081760)         423         408  
9                Lentzea_albida(AB006176)         214         163

id  \
6068  Bacteria;Actinobacteria;Actinobacteria;Actinom...   

                                               sequence  \
6068  ACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGATGAAC...   

                                  species  distance_0  distance_1  
6068  Streptomyces_griseorubens(AB184139)         203          39

id  \
0  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
1  Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   
2  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
3  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   
4  Bacteria;Proteobacteria;Betaproteobacteria;Bur...   
5  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
6  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   
7  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
8  Bacteria;Proteobacteria;Gammaproteobacteria;Al...   
9  Bacteria;Actinobacteria;Actinobacteria;Actinom...   

                                            sequence  \
0  GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   
1  GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   
2  TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   
3  AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   
4  ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   
5  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...   
6  ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...   
7  TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...   
8  AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...   
9  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   

                                  species  distance_0  distance_1  distance_2  
0  Mycobacterium_heidelbergense(AJ000684)         244         228         370  
1             Vibrio_atlanticus(EF599163)         447         411         305  
2      Mycobacterium_aubagnense(AY859683)         239         226         369  
3      Acetobacter_senegalensis(AY883036)         458         435         408  
4    Aquincola_tertiaricarbonis(DQ656489)         420         409         283  
5         Couchioplanes_caeruleus(D85479)         219         182         358  
6   Maritimibacter_alkaliphilus(DQ915443)         433         400         405  
7        Nocardioides_fonticola(EF626689)         273         256         401  
8       Shewanella_schlegeliana(AB081760)         423         408         292  
9                Lentzea_albida(AB006176)         214         163         343

id  \
3661  Bacteria;Proteobacteria;Gammaproteobacteria;Xa...   

                                               sequence  \
3661  GAGTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGCAG...   

                                species  distance_0  distance_1  distance_2  
3661  Lysobacter_panaciterrae(AB245359)         394         376          88

In [8]:
# First, check if there's an existing 'distance' column and drop it
if 'distance' in df1.columns:
    df1.drop(columns=['distance'], inplace=True)

# Now calculate the Levenshtein distance for each sequence
string_to_compare_0 = seq_list[0]
df1['distance_0'] = df1['sequence'].apply(distance, args=(string_to_compare_0,))

string_to_compare_1 = seq_list[1]
df1['distance_1'] = df1['sequence'].apply(distance, args=(string_to_compare_1,))

string_to_compare_2 = seq_list[2]
df1['distance_2'] = df1['sequence'].apply(distance, args=(string_to_compare_2,))

# Display the DataFrame with the updated distance columns
display(df1.head(10))

# Calculate the minimum distances and find the best matches
min_distance_0 = df1['distance_0'].min()
best_match_0 = df1.query("distance_0 == @min_distance_0")
display(best_match_0)

min_distance_1 = df1['distance_1'].min()
best_match_1 = df1.query("distance_1 == @min_distance_1")
display(best_match_1)

min_distance_2 = df1['distance_2'].min()
best_match_2 = df1.query("distance_2 == @min_distance_2")
display(best_match_2)

id  \
0  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
1  Bacteria;Proteobacteria;Gammaproteobacteria;Vi...   
2  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
3  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   
4  Bacteria;Proteobacteria;Betaproteobacteria;Bur...   
5  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
6  Bacteria;Proteobacteria;Alphaproteobacteria;Rh...   
7  Bacteria;Actinobacteria;Actinobacteria;Actinom...   
8  Bacteria;Proteobacteria;Gammaproteobacteria;Al...   
9  Bacteria;Actinobacteria;Actinobacteria;Actinom...   

                                            sequence  \
0  GAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGGAAAGGTC...   
1  GTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   
2  TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAAC...   
3  AAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAAC...   
4  ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGTAAC...   
5  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAG...   
6  ATGGCTCAGAACGAACGCTGGCGGCAGGCCTAACACATGCAAGTCG...   
7  TCTGTAAGTAGCAGGAAAGGCCCTTCGGGGTACTCGAGCGGCGAAC...   
8  AGGCCTAACACATGCAAGTCGAGCGGTAACACAAGGGAGCTTGCTC...   
9  GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAG...   

                                  species  distance_0  distance_1  distance_2  
0  Mycobacterium_heidelbergense(AJ000684)         244         228         370  
1             Vibrio_atlanticus(EF599163)         447         411         305  
2      Mycobacterium_aubagnense(AY859683)         239         226         369  
3      Acetobacter_senegalensis(AY883036)         458         435         408  
4    Aquincola_tertiaricarbonis(DQ656489)         420         409         283  
5         Couchioplanes_caeruleus(D85479)         219         182         358  
6   Maritimibacter_alkaliphilus(DQ915443)         433         400         405  
7        Nocardioides_fonticola(EF626689)         273         256         401  
8       Shewanella_schlegeliana(AB081760)         423         408         292  
9                Lentzea_albida(AB006176)         214         163         343

id  \
8329  Bacteria;Actinobacteria;Actinobacteria;Actinom...   

                                               sequence  \
8329  TGGCGGCGTGCTTAACACATGCAAGTCGAACGGTGAACACGGAGCT...   

                             species  distance_0  distance_1  distance_2  
8329  Microbacterium_oxydans(Y17227)         140         202         349

id  \
6068  Bacteria;Actinobacteria;Actinobacteria;Actinom...   

                                               sequence  \
6068  ACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAACGATGAAC...   

                                  species  distance_0  distance_1  distance_2  
6068  Streptomyces_griseorubens(AB184139)         203          39         347

id  \
3661  Bacteria;Proteobacteria;Gammaproteobacteria;Xa...   

                                               sequence  \
3661  GAGTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAACGGCAG...   

                                species  distance_0  distance_1  distance_2  
3661  Lysobacter_panaciterrae(AB245359)         394         376          88

## Commentary

Three colonies—colony 15, colony 16 and colony C6—were selected for sequencing since they show inhibition against pathogens in overlay test. For colony 15, the closest species match is Streptomyces griseorubens (AB184139), with Levenshtein distance of 39. The closest match to colony 16 is Microbacterium oxydans (Y17227), having a distance of 140. Lastly, Lysobacter panaciterrae (AB245329) is the closest match to colony C6, with a Levenshtein distance of 88.
